src/rust/catnap/win/winsock.rs

```diff
@@ -0,0 +1,348 @@
+        // RIO_EXTENSION_FUNCTION_TABLE.
+        unsafe {
+            WinsockRuntime::do_ioctl(
+                s,
+                SIO_GET_MULTIPLE_EXTENSION_FUNCTION_POINTER,
+                Some(&WSAID_MULTIPLE_RIO),
+                Some(&mut result),
+            )
+        }?;
+
+        if result.cbSize != std::mem::size_of::<RIO_EXTENSION_FUNCTION_TABLE>() as u32 {
+            Err(Fail::new(
+                libc::EFAULT,
+                "Winsock did not return enough data for RIO_EXTENSION_FUNCTION_TABLE",
+            ))
+        } else {
+            Ok(result)
+        }
+    }
+
+    /// Lookup a single function pointer using SIO_GET_EXTENSION_FUNCTION_POINTER ioctl. To be sound, T must be a `fn`
+    /// type.
+    fn lookup_single_fn<T>(s: SOCKET, guid: &GUID) -> Result<Option<T>, Fail> {
+        let mut fn_ptr: Option<T> = None;
+
+        // Safety: SIO_GET_EXTENSION_FUNCTION_POINTER expects type GUID for input. Option<fn()> is compatible with C
+        // output type for this ioctl.
+        unsafe {
+            WinsockRuntime::do_ioctl(s, SIO_GET_EXTENSION_FUNCTION_POINTER, Some(guid), Some(&mut fn_ptr)).map_err(
+                |err| {
+                    let msg: String = format!("{} for function lookup {:?}", err.cause, guid);
+                    Fail::new(err.errno, msg.as_str())
+                },
+            )?;
+        }
+
+        if fn_ptr.is_some() {
+            Ok(fn_ptr)
+        } else {
+            Err(Fail::new(libc::ENOTSUP, "Winsock extension not supported"))
+        }
+    }
+}
+
+impl WinsockRuntime {
+    /// Start up the Winsock runtime, creating a new instance of WinsockRuntime. While it is not functionally useful,
+    /// it is valid to instantiate multiple instances of this struct.
+    pub fn new() -> Result<Self, Fail> {
+        let mut data: WSADATA = WSADATA::default();
+        if unsafe { WSAStartup(0x202u16, &mut data as *mut WSADATA) } != 0 {
+            return Err(expect_last_wsa_error());
+        }
+
+        Ok(WinsockRuntime {
+            extensions_by_provider: HashMap::new(),
+        })
+    }
+
+    /// Implementation of `ioctl`
+    pub(super) unsafe fn do_ioctl<T, U>(
+        s: SOCKET,
+        control_code: u32,
+        input: Option<&T>,
+        output: Option<&mut U>,
+    ) -> Result<(), Fail> {
+        let input: Option<*const libc::c_void> = input.map(|t: &T| -> *const libc::c_void { (t as *const T).cast() });
+        let input_size: usize = input.map(|_| std::mem::size_of::<T>()).unwrap_or(0);
+
+        let output: Option<*mut libc::c_void> = output.map(|u: &mut U| -> *mut libc::c_void { (u as *mut U).cast() });
+        let output_size: usize = output.map(|_| std::mem::size_of::<U>()).unwrap_or(0);
+
+        if input_size > u32::MAX as usize || output_size > u32::MAX as usize {
+            return Err(Fail::new(libc::E2BIG, "WSAIoctl parameter is too big"));
```

Comment:
Would it be helpful to know which one is too big - for debugging? (input/output)